src/Microsoft.ML.TensorFlow/TensorFlow/TensorflowUtils.cs

```diff
@@ -158,6 +160,101 @@ internal static TFSession LoadTFSession(IExceptionContext ectx, byte[] modelByte
+        {
+            FileAttributes attr = File.GetAttributes(modelPath);
+            return attr.HasFlag(FileAttributes.Directory);
+        }
+
+        internal static void CreateTempDirectory(string tempDirPath)
+        {
+            //if directory exists, do nothing.
+            if (Directory.Exists(tempDirPath))
+                return;
+
+            WindowsIdentity currentIdentity = null;
+            try
+            {
+                currentIdentity = WindowsIdentity.GetCurrent();
+            }
+            catch (PlatformNotSupportedException)
+            { }
+
+            if (currentIdentity != null && new WindowsPrincipal(currentIdentity).IsInRole(WindowsBuiltInRole.Administrator))
+            {
+                // Create high integrity dir and set no delete policy for all files under the directory.
+                // In case of failure, throw exception.
+                CreateTempDirectoryWithAcl(tempDirPath, currentIdentity.User.ToString());
+            }
+            else
+                Directory.CreateDirectory(tempDirPath);
+        }
+
+        private static void CreateTempDirectoryWithAcl(string dirPath, string identity)
+        {
+            // Dacl Sddl string:
+            // D: Dacl type
+            // D; Deny access
+            // OI; Object inherit ace
+            // SD; Standard delete function
+            // wIdentity.User Sid of the given user.
+            // A; Allow access
+            // OICI; Object inherit, container inherit
+            // FA File access
+            // BA Built-in administrators
+            // S: Sacl type
+            // ML;; Mandatory Label
+            // NW;;; No write policy
+            // HI High integrity processes only
+            string sddl = "D:(D;OI;SD;;;" + identity + ")(A;OICI;FA;;;BA)S:(ML;OI;NW;;;HI)";
+
+            var dir = Directory.CreateDirectory(dirPath);
+            DirectorySecurity dirSec = new DirectorySecurity();
+            dirSec.SetSecurityDescriptorSddlForm(sddl);
+            dirSec.SetAccessRuleProtection(true, false);  // disable inheritance
+            dir.SetAccessControl(dirSec);
+
+            // Cleaning out the directory, in case someone managed to sneak in between creation and setting ACL.
+            DirectoryInfo dirInfo = new DirectoryInfo(dirPath);
+            foreach (FileInfo file in dirInfo.GetFiles())
+            {
+                file.Delete();
+            }
+            foreach (DirectoryInfo subDirInfo in dirInfo.GetDirectories())
+            {
+                subDirInfo.Delete(true);
+            }
+        }
+
+        internal static TFSession GetSession(IHostEnvironment env, string modelPath)
+        {
+            if (IsSavedModel(modelPath))
+                return LoadTFSession(env, modelPath);
+
+            return CheckFileAndRead(env, modelPath);
+        }
+
+        private static TFSession CheckFileAndRead(IHostEnvironment env, string modelFile)
```

Comment:
>private static TFSession CheckFileAndRead(IHostEnvironment env, string modelFile) [](start = 8, length = 81)

can you get rid of this method? no one using it any more except GetSession, and I think it would be much better to move checks in start of GetSesssion